File: 1092-maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    void solve(TreeNode* root,int maxi,int mini){
        if(!root) return;
        maxi=max(maxi,root->val);
        mini=min(mini,root->val);
        if(root->left) solve(root->left,maxi,mini);
        if(root->right) solve(root->right,maxi,mini);
        ans=max(ans,(maxi-mini));
    }
    int maxAncestorDiff(TreeNode* root) {
        solve(root,0,INT_MAX);
        return ans;
    }
};

File: 1092-maximum-difference-between-node-and-ancestor/maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    ans=0
    def solve(self,root,maxi,mini):
        if not root:
            return
        maxi=max(maxi,root.val)
        mini=min(mini,root.val)
        if(root.left):
            self.solve(root.left,maxi,mini)
        if(root.right):
            self.solve(root.right,maxi,mini)
        self.ans=max(self.ans,maxi-mini)
    def maxAncestorDiff(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        self.solve(root,0,100000)
        return self.ans
        